src/osd/scheduler/mClockScheduler.cc

```diff
@@ -51,32 +51,52 @@ mClockScheduler::mClockScheduler(CephContext *cct,
 {
   cct->_conf.add_observer(this);
   ceph_assert(num_shards > 0);
-  set_max_osd_capacity();
-  set_osd_mclock_cost_per_io();
-  set_osd_mclock_cost_per_byte();
+  set_max_osd_random_write_iops();
+  set_max_osd_sequential_bandwidth();
+  set_osd_bandwidth_cost_per_io();
   set_mclock_profile();
   enable_mclock_profile_settings();
-  client_registry.update_from_config(cct->_conf);
+  client_registry.update_from_config(cct->_conf, osd_bandwidth_cost_per_io);
 }
 
-void mClockScheduler::ClientRegistry::update_from_config(const ConfigProxy &conf)
+/* Update client infos for various classes of clients within the mClock server.
+ *
+ * Convert the allocations for each client class from IOPS per shard to
+ * bandwidth per shard using osd_bandwidth_cost_per_io as shown below:
+ *
+ *   Bandwidth/shard = osd_bandwidth_cost_per_io * IOPS/shard
+ *    (Bytes/sec)           (Bytes/io)                (iops)
+ *
+ * Update the default external and internal client infos in the mClock server
+ * with allocations in Bytes/sec. Eventually, these allocations are converted to
+ * secs by the mClock server as part of tag calculations.
+ */
+void mClockScheduler::ClientRegistry::update_from_config(
+  const ConfigProxy &conf,
+  const double cost_per_io)
 {
```

Comment:
What units do osd_mclock_scheduler_client_res, osd_mclock_scheduler_client_wgt, and osd_mclock_scheduler_client_lim have?  Are they supposed to be pure ratios?  From osd.yaml.in:

```
- name: osd_mclock_scheduler_client_res
...
  desc: IO proportion reserved for each client (default)
  long_desc: Only considered for osd_op_queue = mclock_scheduler
  fmt_desc: IO proportion reserved for each client (default).
  default: 1
...
- name: osd_mclock_scheduler_client_wgt
...
  desc: IO share for each client (default) over reservation
  long_desc: Only considered for osd_op_queue = mclock_scheduler
  fmt_desc: IO share for each client (default) over reservation.
  default: 1
...
- name: osd_mclock_scheduler_client_lim
...
  desc: IO limit for each client (default) over reservation
  long_desc: Only considered for osd_op_queue = mclock_scheduler
  fmt_desc: IO limit for each client (default) over reservation.
  default: 999999
...
```

From the above:
- osd_mclock_scheduler_client_res should be a "proportion" and thus a pure ratio
- osd_mclock_scheduler_client_wgt should be a "share" and thus a pure ratio
- osd_mclock_scheduler_client_lim is a "limit".  From the default (999999) I assume it is meant to be in cost and therefore bytes?

Below, we are invoking:

```
  default_external_client_info.update(
    cost_per_io * conf.get_val<uint64_t>("osd_mclock_scheduler_client_res"),
    conf.get_val<uint64_t>("osd_mclock_scheduler_client_wgt"),
    cost_per_io * conf.get_val<uint64_t>("osd_mclock_scheduler_client_lim"));
```

in mClockScheduler::ClientRegistry::update_from_config.  cost_per_io has units bytes/io.  Thus, we are passing the following into update:

- cost_per_io (bytes/io) * osd_mclock_scheduler_client_res (pure ratio) = bytes/io
- osd_mclock_scheduler_client_wgt (ratio) = pure ratio
- cost_per_io (bytes/io) * osd_mclock_scheduler_client_lim (bytes) = bytes^2/io

Several things seem to be off here.

First, the defaults for osd_mclock_scheduler_client_res and osd_mclock_scheduler_client_wgt are both 1, which doesn't make sense for a pure ratio (unless the default is for clients to have a reservation of all of the OSD's capacity?).  What's the intent here?

The first parameter to  default_external_client_info.update appears to have units bytes/io.  My understanding is that the reservation parameter for mclock should have units of cost/time as it represents the amount of cost per unit time the osd will reserve for clients.  Presumably this should be ```max_osd_sequential_bandwidth (bytes/sec) * osd_mclock_scheduler_client_res```?  See above my comment on the representationa and default, though.

I think the second parameter seems correct (except that the default of 1 doesn't really make sense to me) assuming that the weight param in mclock is supposed to be a pure ratio representing the proportion of spare capacity that should be given to clients.  Does weight get normalized elsewhere?

The last parameter is the limit param which should have units of bytes/second.  I think osd_mclock_scheduler_client_lim should probably be modified to be a pure ratio representing the max proportion of the OSD allowed to be used for a client.  Then, the last param should probably be ```max_osd_sequential_bandwidth (bytes/sec) * osd_mclock_scheduler_client_lim```

Reply:
How does update_from_config relate to ```set_*_profile_allocations```?  It seems like those three override op_scheduler_class::background_recovery and op_scheduler_class::background_best_effort set here, set op_scheduler_class::client additionally, and leave default_external_client_info.  Is the idea that we use ```osd_mclock_scheduler_\*_(res|lim|wgt)``` if no profile is specified?